[Messenger] Worker events + global retry functionality

weaverryan opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes, on Messenger only
| Deprecations? | no
| Tests pass?   | NEEDED
| Fixed tickets | #29132, #27008, #27215 and part of #30540
| License       | MIT
| Doc PR        | TODO

This is an alternative to #29132 and #27008. There are several big things:

1) The `messenger:consume` does not die if a handler has an error
2) Events are dispatched before, after and on error a message being handled
3) Logic is moved out of Amqp and into the Worker so that we can have some consistent features, like error handling.
4) A generic retry system was added, which works with Amqp and future transports should support. 
 It will work out of the box for users. Retrying works by putting the received `Envelope` back into the bus, but with the `ReceivedStamp` removed. The retry functionality has an integration test for AMQP.
5) Added a new `MessageDecodingFailedException` that transport Serializers should throw if `decode()` fails. It allows us to reject a message in this situation, as allowing it to fail but remain on the queue causes it to be retried forever.
6) A new `DelayStamp` was added, which is the first of (later) more stamps for configuring the transport layer (see #30558).

BC breaks are documented in the CHANGELOG.

Thanks!

weaverryan committed:
Adding global retry support, events & more to messenger transport

Co-authored-by: Samuel ROZE <samuel.roze@gmail.com>

weaverryan commented:
I've now incorporated generic retry abilities from #27008. I think more details need to be considered, like #30558 (transport stamp) and how it will affect things.

But, generally speaking, what big issues do people see? I'm pushing much more control into the `Worker`, and making the transport a bit more agnostic to sending/receiving/acking/nacking/retrying.

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php:
```diff
@@ -42,51 +42,72 @@ public function __construct(Connection $connection, SerializerInterface $seriali
+            $handler($this->serializer->decode([
+                'body' => $AMQPEnvelope->getBody(),
+                'headers' => $AMQPEnvelope->getHeaders(),
+            ]), new QueuedMessageMetadata(
```
Can't this `QueuedMessageMetadata` simply be a Stamp added to the decoded envelope? It's very much related to retries so I'd make it explicit in the name.

weaverryan replied:
I've changed this to a stamp indeed! Thought, the stamp now only carries the current retry number, the "message" is carried via "state" inside the receiver now... which I did try avoid leaking this internal object externally on a stamp.

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php:
```diff
@@ -42,51 +42,72 @@ public function __construct(Connection $connection, SerializerInterface $seriali
         }
     }
 
+    public function acknowledge($message): void
```
IMHO these methods should take the envelope, so it's consistent with the other ones and could even be moved to an interface later.

weaverryan replied:
I've removed this arg internally, and they now rely on a "currentAmqpMessage". It's temporary state... but I think it's safe, and it keeps that internal `$message` "thing" (which will vary based on the transport) inside the transport.

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php:
```diff
@@ -24,6 +24,8 @@
  */
 class Connection
 {
+    public const ATTEMPT_COUNT_HEADER_NAME = 'symfony-messenger-attempts';
```
AMQP uses `x-death` when a message is dead-letter-ed. Maybe we can reuse the same? (not sure if it conflicts though, would need to try).

weaverryan replied:
Hmm. I don't know much about this, but I don't think it'll work. Each time we retry, we're re-using the headers from the existing message. This has a nice effect that the `x-death` header (which is an array) will have 0 items, then 1 item, then 2 items, etc - it'll increase with each "death". The "newest" death apparently (i've just tested) always becomes the `0` key - the others are "pushed back". If we try to insert an entry into `x-death`, it just looks like there was a "previous" death, and Rabbit pushes a new item onto the 0 index and our entry is pushed back to 0.

It *seems* like we shouldn't be setting values onto `x-death`.

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php:
```diff
@@ -42,51 +42,72 @@ public function __construct(Connection $connection, SerializerInterface $seriali
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
+            $handler($this->serializer->decode([
```
You still need to catch the `RejectMessageExceptionInterface` exception and reject, isn't it? It's removing a very convenient feature here. 🤔

weaverryan replied:
It's still there as a `UnrecoverableMessageHandlingException`. I had to move it because it's now outside of the `Amqp` namespace, and I also renamed it. But I can rename it back (but it should be in a new namespace either way)

sroze commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -21,11 +23,42 @@ interface ReceiverInterface
+    /**
+     * Acknowlege that the message was handled (i.e. remove from queue).
+     *
+     * @param mixed $message A message as understood by the transport
```
As mentioned, I'd go for the `Envelope` here.

sroze commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -21,11 +23,42 @@ interface ReceiverInterface
+     *
+     * @throws TransportException If there is an issue communicating with the transport
+     */
+    public function acknowledge($message): void;
```
Can't we have these methods in another interface instead? Because it doesn't apply to all of them so I'm not sure we should put them here.

weaverryan replied:
You're thinking like when a receiver is an API endpoint? Hmm. What might you call that? It's like a `QueueReceiverInterface`??? I'm interested in this...

sroze replied:
`AsynchronousReceiverInterface`?

weaverryan replied:
Actually, it's done! It's not exactly another interface, `ReceiverInterface` now has a `getResponder()` method, which returns an optional object that's able to "respond" back to the queue with ack/nack sort of things.

The reason this couldn't live in a totally independent interface (e.g. so that `AmqpSender implements SenderInterface, QueueReceiverInterface`) is that it killed decoration around `SenderInterface`, like `StopWhenMemoryUsageIsExceededReceiver` because these would only decorate `SenderInterface`... but maybe the sender actually implemented `QueueReceiverInterface` also... hence the, sort of, odd solution.

sroze commented on src/Symfony/Component/Messenger/Worker.php:
```diff
@@ -43,12 +56,63 @@ public function run()
+        }
+
+        $numberOfRetries = $messageMetadata->getNumberOfTimesRetried() + 1;
+        // TODO - make retry attempts configurable.
```
I'd suggest they should be configured via the transport DSN. Obviously you wouldn't have it anymore but you could have it as part of the `QueuedMessageMetadata` (to be moved in an envelope stamp) that could have a `getRemainingAttempts()` method or similar.

weaverryan replied:
I refactored this as well. I now use a `RetryStrategyInterface` (with a default implementation) that can be configured on a transport-by-transport basis. The logic is actually handled outside of the transport, but the user can configure on a transport-by-transport basis

weaverryan commented:
Thank you @sroze  for the review! I was working concurrently with your review, so I pushed a bunch of changes as you were commenting. I would love another look. I've replied to a few of your comments with the changes I've made. Specifically, the `QueuedMetadata` thing is now gone, replaced with a Stamp. Also (and this is important), I've given the `AmqpReceiver` some temporary state. The issue is that the `AMQPEnvelope` is needed to do things like `ack`, `nack`, etc. And because these methods are now called by `Worker`, I previously was passing this "message" back out to the Worker... which was unfortunate, because this is really an internal detail to each transport. I fixed that by tracking the current `AMQPEnvelope` being handled inside `AmqpReceiver`, which allows us to drop passing the `AMQPEnvelope` object around.

I also made the retry stuff configurable on a transport-by-transport basis using a service.

Thanks!

weaverryan commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -21,11 +23,40 @@ interface ReceiverInterface
+     *
+     * @throws TransportException If there is an issue communicating with the transport
+     */
+    public function retryCurrentMessage(int $delay): void;
```
These all rely on being called from inside the `receive()` loop, because that keeps temporary "state" about which message is currently being handled. That's a key change here, which simplifies a lot, but which I want to make sure won't cause issues.

weaverryan commented:
This is ready for review! This represents a big change in how we handle the transports, so I really appreciate review!

weaverryan commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php:
```diff
@@ -55,40 +64,70 @@ public function receive(callable $handler): void
+                    'body' => $this->currentAmqpEnvelope->getBody(),
+                    'headers' => $this->currentAmqpEnvelope->getHeaders(),
+                ]);
+            } catch (MessageDecodingFailedException $exception) {
```
This is a responsibility that we *could* move into `Worker`, so that it's not something that the individual receivers needs to worry about.

sroze replied:
I agree 👍

weaverryan commented on src/Symfony/Component/Messenger/Worker.php:
```diff
@@ -48,7 +65,81 @@ public function run()
+                'class' => \get_class($message),
+            ];
+
+            try {
```
There is a try-catch around the bus, which means if an exception is thrown there, we're handled.

However, there is currently no try-catch around this *whole* thing - around the *whole* `$this->receiver->receive` call. This means that any exceptions thrown by the receiver while reading/decoding the message (specifically `TransportException` or a decoding exception) are not caught. 

Should we catch those and, for example, dispatch another exception? Or should we let that exception be throw, which would naturally trigger the console exception event, and cause the `messenger:consume` command to exit? Exiting is not the end-of-the-world, as the user will need something keep workers running anyways (e.g. supervisord)

sroze replied:
I'd let it throw, I'm not sure what would we do otherwise anyway?

weaverryan commented on src/Symfony/Component/Messenger/Worker.php:
```diff
@@ -48,7 +65,81 @@ public function run()
+                $this->dispatchEvent(
+                    WorkerMessageFailedEvent::class,
+                    new WorkerMessageFailedEvent($envelope, $this->receiverName, $throwable, $shouldRequeue)
+                );
```
The events are currently dispatched before ack/reject/retry on the queue. That's subjective, and either order could, in theory, cause a situation where one fails and so the other doesn't run (e.g. some listener throws an exception, so the retry never happens, or, if we reverse, the retry fails due to a network connection, then the event is never dispatched). Not sure if we need to be thinking about this level of failure. Catching exceptions makes things harder to debug/know about when they go wrong.

sroze commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1098,6 +1098,18 @@ function ($a) {
+                                        ->end()
+                                        ->children()
+                                            ->scalarNode('service')->defaultValue('messenger.retry.exponential_retry_strategy')->end()
+                                            ->scalarNode('max_retries')->defaultValue(3)->end()
```
Wouldn't it make more sense for these options to be part of the `dsn`?
```
DSN=amqp://...?retry[max_retry]=3&retry[exponential_offset]=0
```

weaverryan replied:
It's not logic that ultimately ends up in the transport. It more ties to a "global retry system" that you can configure on a transport-by-transport basis.

And on a personal note, I'd like to keep things out of the DSN, unless it's really transport-specific connection config. One up-side is that this retry logic can be committed.

javiereguiluz replied:
Should `max_retries` be `scalarNode()` or `integerNode()`? If we change it to integer, should we add `->min(0)` too?

sroze commented on src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php:
```diff
@@ -140,8 +145,13 @@ protected function execute(InputInterface $input, OutputInterface $output): void
             throw new RuntimeException(sprintf('Bus "%s" does not exist.', $busName));
         }
 
+        if (!$this->retryStrategyLocator->has($receiverName)) {
```
Shouldn't we decorate the receiver instead of having this logic in the command & worker then? 🤔

weaverryan replied:
We probably could, but *should* we? The code reads really clearly inside `Worker`, and my thought is sort of that we're setting out the "core" logic that (unless you really want to) everyone gets. At this point, it includes a lot - event dispatching, retry logic (and a stamp being added for this).

So I guess I would say: someone needs to sell hard on the idea of making it a decorator.

weaverryan replied:
Or, to say it differently:

* Do we *really* need this?
* If we do want it, could we only move some parts out (e.g. logging decorator, or event listener, event decorator)  and leave others

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php:
```diff
@@ -42,11 +46,16 @@ public function __construct(Connection $connection, SerializerInterface $seriali
-            $AMQPEnvelope = $this->connection->get();
-            if (null === $AMQPEnvelope) {
+            try {
+                $this->currentAmqpEnvelope = $this->connection->get();
```
I'm not fond of the idea of forcing the receivers to have such local state at all. Can't you have another stamp (non-serialisable), AMQP-specific, that contains this `\AMQPMessage`? You'd call it `AmqpStamp` and that's it, you can just get it from the `Envelope`, no need of these `*currentMessage` methods anymore, just "normal" methods taking the `Envelope` as an argument.

weaverryan replied:
Indeed, I've gone back-and-forth on this. Your transport-specific stamp is a nice idea. It *would* leak this information out to userland, like middleware or event listeners. What do you think about that?

Also, in the latest commit, on retry, I re-send the Envelope for normal encoding/decoding. This means (and is by design) that if a new Envelope was created with new stamps, those stamps will be serialized & sent. It would mean that this AMQP-specific stamp *would* be serialized & sent. You mentioned "non-serialisable"... do you basically mean: give it a "sleep" method so that if/when we serialize it, it'll just be an empty object (i.e. it won't cause an error).

weaverryan replied:
Side note: if we did this and made methods like `reject()` require the `Envelope`, we wouldn't be able to handle the `MessageDecodingFailedException` in the `Worker` - catching that and rejecting would need to remain the responsibility of each transport, because there wouldn't be any `Envelope` that the `Worker` could send back to the `Receiver::reject()` method.

sroze replied:
> It would leak this information out to userland, like middleware or event listeners. What do you think about that?

That's a good thing IMHO. This allows users to go deep into customising it.

> This means (and is by design) that if a new Envelope was created with new stamps, those stamps will be serialized & sent.

Indeed, hence my (non-serializable) comment. We need a way to make them non-transportable, it's definitely a valid use-case. The `sleep` method works but is only about the `serialize` method... I'd imagine a `NonSerializableStampInterface` actually.

> if we did this and made methods like reject() require the Envelope, we wouldn't be able to handle the MessageDecodingFailedException in the Worker

Technically speaking we could actually create an empty envelope as part of the exception. It would make sense. But for the sake of this pull-request, let's keep it the responsibility of the transport :)

weaverryan replied:
I've made the change to use a stamp (`AmqpReceivedStamp`) and it's really, really nice - good suggestion. However, I also found out that serializing that stamp works just fine. The `AMQPEnvelope` inside it is just a simple object that doesn't cause any problems.

So, should we still add this `NonSerializableStampInterface`? Serializing the `AMQPEnvelope` doesn't cause any problems, as we're always looking for `$envelope->last(AMQPEnvelop::class)`. Also, to implement this, each serializer would actually need to create a *new* envelope with the `NonSerializableStampInterface` filtered out. Totally doable, but is this something we actually need to add?

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php:
```diff
@@ -55,40 +64,70 @@ public function receive(callable $handler): void
+            );
+
+            if (\function_exists('pcntl_signal_dispatch')) {
+                pcntl_signal_dispatch();
```
This is done by the `Worker` as well. Do we really need it here actually? :)

weaverryan replied:
The `Worker` does not do this currently, but, should it perhaps? I'm all for moving more boilerplate to the `Worker` so that transports don't need to bother with it.

Nyholm replied:
:+1: for having this in the worker

weaverryan replied:
Moved into the Worker

sroze replied:
You can remove it from here now then :)

weaverryan commented:
One last update: "retrying" is no longer a special situation - we basically just call `SenderInterface::send()` with whatever the delay is, then "ack" the old message. That is *basically* what we were doing before, but the difference is that we're now re-serializing the Envelope on re-send, instead of sending a duplicate of the original message. That allows us to add stamps to the Envelope, which is *really* powerful because we can, for example, manage the `RetryCountStamp` in a way where the transport doesn't need to do *anything* with tracking how many retries have happened.

Guikingone commented on src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php:
```diff
@@ -196,8 +307,8 @@ public function setup(): void
     public function channel(): \AMQPChannel
     {
         if (null === $this->amqpChannel) {
```
Can be checked via `\is_null()` for opcode optimisations :)

nicolas-grekas replied:
Nope sorry, that's not our CS, and that wouldn't change anything actually :)

nicolas-grekas commented on src/Symfony/Component/Messenger/CHANGELOG.md:
```diff
@@ -3,7 +3,28 @@ CHANGELOG
+   3 additional optional arguments.
+ * [BC BREAK] `ConsumeMessagesCommand` has a new, required third argument.
+ * [BC BREAK] `RejectMessageExceptionInterface` was removed and replaced
+   by `Symfony\Component\Messenger\Exception]UnrecoverableMessageHandlingException`,
```
`s/]/\\/`

Nyholm commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -1657,6 +1658,13 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
+
+            $transportRetryReferences[$name] = new Reference($transport['retry_strategy']['service']);
+            if ('messenger.retry.exponential_retry_strategy' === $transport['retry_strategy']['service']) {
+                $container->getDefinition('messenger.retry.exponential_retry_strategy')
```
Im not sure I like this. This makes `exponential_retry_strategy` very special... 

- Should we add a list with all our supported strategies here? 
- How do I do if I want some config to my custom strategy?

Or maybe Im overthinking... This is how we configure our symfony services and yes, they should be treated special.

Nyholm replied:
We should validate the configuration so we cant use `max_retries` when we dont use `exponential_retry_strategy`

weaverryan replied:
> We should validate the configuration so we cant use max_retries when we dont use exponential_retry_strategy

Yes! This is already done

> Im not sure I like this. This makes exponential_retry_strategy very special...

I'm also not sure about this. By giving *one* strategy, it made configuration easy. If we also add a second strategy, for example, then we would also need to make its options configurable. So, I took the middle ground of: let's give them one nice strategy out-of-the-box.

> How do I do if I want some config to my custom strategy

Create a class that implements the interface, then set its id on the `retry_strategy.service` under the transport. So, retry strategies are transport-by-transport.

I'd also like to make the retry stuff configurable on a message-by-message basis. Perhaps there could be a stamp that the user adds that defines the retry behavior for a specific message.

weaverryan replied:
I've replaced the exponential retry strategy with a more generic, flexible one that should cover 95% of use-cases. The last 5% can register their own service for this.

Nyholm commented on src/Symfony/Component/Messenger/Retry/ExponentialRetryStrategy.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $retries = $this->getCurrentRetryCount($message) + 1;
+
+        return pow(2, $retries + $this->offset) * 1000;
```
With offset: -2 and retries: 3 
Then you will wait 0.5s, 1s, 2s. 

The default offset: 0 and retries: 3
Then you will wait 2s, 4s, 8s.

weaverryan replied:
Correct. Is that a problem? I wasn't really thinking about negative offsets... but it still seems to work ok.

Nyholm replied:
No. This is fine. At first I thought 2, 4, 8 could be too much of delay.   I just wanted to try different options and I shared the results.

Nyholm commented on src/Symfony/Component/Messenger/Stamp/TransportConfigurationStamp.php:
```diff
@@ -0,0 +1,30 @@
+{
+    private $delay = 0;
+
+    // TODO - add more stuff!
```
And/or move away from type safety and add a generic “data” store

weaverryan replied:
Indeed. I think there are a few things that we should generically support - like delay, priority, maybe ttl. From a DX perspective, I'd like these to have proper setter methods. But then there will also need to be dynamic, transport-specific stuff, like `routingKey`, which would require a generic data store. So... both?

Nyholm replied:
Say getters and setters for “normal” stuff. But we should only use one property. Then one could easily implement “getAll()”

That would be super dynamic for MyCustomTransport.

sroze replied:
I'd vote in favour of more specific stamps actually instead of a generic one. `DelayedStamp`, `PriorityStamp`, etc... it doesn't cost much and makes everything much more explicit.

weaverryan replied:
Yes, I like this - it's even cleaner/easier when you're creating the envelope and setting this stuff on the stamps. I've just added `DelayStamp`

Nyholm reviewed (commented):
I did a review on the web. I will have to check out the code to understand the architecture and the big picture.

weaverryan commented:
2 big architectural things:

1) the new method on ReceiverInterface is a compromise between adding 3 new methods to that interface that not all need versus giving us the functionality we need. I’d like thoughts on that decision.

2) The senders (check out AmqpSender) currently have some temporary “state” where they keep track of the message being processed so they know which message to acknowledge or reject. An alternative would be for each Sender to attach a stamp to the Envelope that says which message is being processed. The downside is that individual senders would need to catch the MessageDecodingFailedException on their own (it’s actually currently done like this) versus doing that in the Worker, because when deciding fails, there’s no Envelope that is sent to the worker that could then be sent back in reject().

So, the questions are:

1) should we add just one new method to ReceiverInterface (as it’s done now) or is there a better way for a Receiver to support the Worker calling acknowledge/reject/send on it?

2) should the receiver keep track of which message it’s processing internally or push that into a stamp?

weaverryan commented on src/Symfony/Component/Messenger/Event/WorkerMessageFailedEvent.php:
```diff
@@ -0,0 +1,42 @@
+        return $this->throwable;
+    }
+
+    public function wasRequeued(): bool
```
Needs to be named willRequeue as the event is currently dispatched before the requeue actually happens

sroze commented on src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpSenderTrait.php:
```diff
@@ -0,0 +1,51 @@
+/**
+ * @internal
+ */
+trait AmqpSenderTrait
```
Instead, shouldn't we merge the receiver and the sender into a single class?

sroze commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverResponderInterface.php:
```diff
@@ -0,0 +1,35 @@
+     *
+     * @throws TransportException If there is an issue communicating with the transport
+     */
+    public function acknowledgeCurrentMessage(): void;
```
```suggestion
    public function acknowledge(Envelope $envelope): void;
```

😉

sroze commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -21,13 +24,29 @@ interface ReceiverInterface
+     * rejected on the responder when inside the receive handler
+     * callback.
+     */
+    public function getResponder(): ?ReceiverResponderInterface;
```
This is only used in the `Worker`. Instead, can you require to be injected the `TransportInterface`, get the sender and check whether it implements `ReceiverResponderInterface`? This way, we don't break this interface (which is implemented by a lot of transports already) as we don't need.

weaverryan replied:
I like your thinking. Oh, bit did you mean "get the sender" or "get the receiver" above? I want to make sure I wasn't misunderstanding that part.

However, as much as I like this, there are two problems:

1) If we inject the `TransportInterface` into `Worker` instead of the `ReceiverInterface`, the decorator classes (like `StopWhenTimeLimitIsReachedReceiver`) would need to be updated to decorate the whole transport).

2) The `TransportInterface` does not have a `getReceiver()` method. So, if we inject the transport into `Worker`, we can't get the receiver without adding this method to the interface (which would still break existing implementations).

Thoughts?

weaverryan replied:
After talking with Nicolas, I've updated this (again). The "responder" is gone (yay!), `ReceiverInterface` has `acknowledgeMessage()` and `rejectMessage()` (because any "decent" receivers will have this, others can leave them blank) and for "retry", we're actually sending the Envelope back through the bus in order to re-send (Nicolas' stroke of genius). We simply remove the `ReceivedStamp` and pass it back in.

sroze commented on src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php:
```diff
@@ -42,4 +43,35 @@ public function encode(Envelope $envelope): array
+     */
+    public static function handleUnserializeCallback($class)
+    {
+        throw new MessageDecodingFailedException('Message class not found during decoding: '.$class);
```
```suggestion
        throw new MessageDecodingFailedException(sprintf('Message class "%s" not found during decoding.', $class));
```

sroze reviewed (changes requested):
Getting there 🤗

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1101,6 +1101,18 @@ function ($a) {
+                                        ->children()
+                                            ->scalarNode('service')->defaultValue('messenger.retry.exponential_retry_strategy')->end()
+                                            ->scalarNode('max_retries')->defaultValue(3)->end()
+                                            ->scalarNode('exponential_offset')->defaultValue(0)->end()
```
Here also we should validate that `exponential_offset` is numeric (and >= 0 ?).

weaverryan replied:
If we keep this option (still checking into if this is the best way to make this all configurable), `exponential_offset` CAN be negative, actually.

javiereguiluz commented on src/Symfony/Component/Messenger/Stamp/RetryCountStamp.php:
```diff
@@ -0,0 +1,32 @@
+        $this->currentRetryCount = $currentRetryCount;
+    }
+
+    public function getCurrentRetryCount(): int
```
Minor comment: is it necessary to add the `Current` word here?  "RetryCount" cannot be mistaken with "MaxRetries"  ... and we don't use the generic (and confusing) "retries" anywhere.

javiereguiluz commented on src/Symfony/Component/Messenger/Stamp/TransportConfigurationStamp.php:
```diff
@@ -0,0 +1,30 @@
+        return $this->delay;
+    }
+
+    public function setDelay(int $delay): self
```
We'd need a comment or some hint here to explain that the delay is in milliseconds (if it really is, I don't know).

javiereguiluz commented on src/Symfony/Component/Messenger/Tests/Fixtures/CallbackReceiver.php:
```diff
@@ -22,4 +30,41 @@ public function receive(callable $handler): void
+        return $this;
+    }
+
+    public function acknowledgeCurrentMessage(): void
```
Could we replace `...CurrentMessage()` by `last...Message()`? The second pattern is more commonly used in Symfony, such as `getLastAuthenticationError()` and `getLastUsername()` in AuthUtils.

weaverryan replied:
the current is now gone from this for other reasons, fortunately :)

javiereguiluz commented:
Probably out of the scope for us ... but for reference purposes, [Spring's Retry project](https://github.com/spring-projects/spring-retry) allows to configure the following retry properties (maybe we want to support defining expressions for example?):

| Type | Property | Description
| ---- | -------- | -----------
long | delay | A canonical backoff period.
String | delayExpression | An expression evaluating to the canonical backoff period.
long | maxDelay | The maximimum wait (in milliseconds) between retries.
String | maxDelayExpression | An expression evaluating to the maximimum wait (in milliseconds) between retries.
double | multiplier |If positive, then used as a multiplier for generating the next delay for backoff.
String | multiplierExpression | Evaluates to a value used as a multiplier for generating the next delay for backoff.
boolean | random | In the exponential case (multiplier() > 0) set this to true to have the backoff delays randomized, so that the maximum delay is multiplier times the previous delay and the distribution is uniform between the two values.

See https://docs.spring.io/spring-retry/docs/api/current/org/springframework/retry/annotation/Backoff.html

weaverryan commented:
@javiereguiluz I've just made the retry logic basically identical to that - it's much nicer. Excellent suggestion!

weaverryan commented on src/Symfony/Component/Messenger/Envelope.php:
```diff
@@ -54,6 +54,18 @@ public function with(StampInterface ...$stamps): self
+        unset($cloned->stamps[$stampFqcn]);
+
+        return $cloned;
+    }
```
Handy method :). Allows us to remove the `ReceivedStamp`

weaverryan commented:
Ok, I think we might have it :). PR description & CHANGELOG are both fully updated.

Last critical change was this:

1) The "responder" idea was killed (it was ugly)
2) `ReceiverInterface` has 2 new methods: `acknowledgeMessage(Envelope $envelope)` and `rejectMessage(Envelope $envelope)`. This came from talking with @nicolas-grekas - the idea is that any "decent receiver should be able to ack/nack messages. But, of course, if you don't want to or don't have that use-case, leave these message blank, or use them for logging
3) For retrying, we're now sending the `Envelope` back through the Bus. This was Nicolas' idea, and it solves the big problem that, in `Worker`, we only have access to the Receiver, so it was super awkward to try to find a way to effectively use the Sender from inside of there. No more: we simply clone the Envelope, remove the `ReceivedStamp` and send it to the bus.

The PR now has no pending issues or things I'm unhappy with. Please review so we can move forward with this and the PR's it's blocking :).

Thanks!

fabpot commented:
While testing this PR, I realized that the command to consume messages was a bit "verbose":`messenger:consume-messages`. What about renaming it to just `messenger:consume`?

fabpot commented on src/Symfony/Component/Messenger/Worker.php:
```diff
@@ -45,10 +64,118 @@ public function run()
+            return;
+        }
+
+        $this->eventDispatcher->dispatch($eventName, $event);
```
`$event` should now be the first argument.

fabpot commented on src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * Returns a unique identifier for transport receiver this message was received from.
+     *
+     * @return string
```
Such phpdocs can be removed as the method signature has the same information already.

fabpot commented on src/Symfony/Component/Messenger/Exception/MessageDecodingFailedException.php:
```diff
@@ -0,0 +1,21 @@
+ *
+ * @experimental in 4.3
+ */
+class MessageDecodingFailedException extends \Exception implements ExceptionInterface
```
We rarely extend `\Exception` directly. Should it be `InvalidArgumentException` instead?

fabpot commented on src/Symfony/Component/Messenger/Exception/UnrecoverableMessageHandlingException.php:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Indicates an exception while handling a message that will continue to fail.
+ *
+ * If something goes wrong while a handling a message from a transport and the
```
type: while handling?

fabpot commented on src/Symfony/Component/Messenger/Retry/MultiplierRetryStrategy.php:
```diff
@@ -0,0 +1,100 @@
+    private $delayMilliseconds;
+
+    private $multiplier;
+
```
I would remove all the blank lines between property definitions to be consistent with the rest of the code base (same goes for other classes).

fabpot commented on src/Symfony/Component/Messenger/Retry/MultiplierRetryStrategy.php:
```diff
@@ -0,0 +1,100 @@
+        $this->maxRetries = $maxRetries;
+
+        if ($delayMilliseconds < 0) {
+            throw new \InvalidArgumentException(sprintf('Delay must be greater than or equal to zero: "%s" passed', $delayMilliseconds));
```
Missing dots at the end of the exception messages (same for other exception messages).

fabpot commented on src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php:
```diff
@@ -53,9 +53,45 @@ class Connection
+    private $amqpDelayExchange;
+
+    /**
+     * Available options:.
```
extra dot at the end.

That's one case where having a generic `Constructor` as the first line is kind of required (if not, `Available options` is the method description).

fabpot commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -23,11 +27,30 @@ interface ReceiverInterface
      * Note that this envelope can be `null` if the timeout to receive something has expired.
+     *
+     * If the received message cannot be decoded, the message should
+     * be removed from the queue and the MessageDecodingFailedException be thrown
```
missing ending dot.

weaverryan commented:
Comments addressed & several BC breaks were replaced by deprecations (CHANGELOG is updated). 

Last important changes from yesterday are described here: https://github.com/symfony/symfony/pull/30557#issuecomment-475006181

This is ready to go!

nicolas-grekas commented on src/Symfony/Component/Messenger/Retry/MultiplierRetryStrategy.php:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * @param int $maxRetries           The maximum number of time to retry (0 means indefinitely)
+     * @param int $delayMilliseconds    Amount of time to delay (or the initial value when multiplier is used)
+     * @param int $multiplier           Multiplier to apply to the delay each time a retry occurs
```
a float would be better here I think

nicolas-grekas commented on src/Symfony/Component/Messenger/Retry/MultiplierRetryStrategy.php:
```diff
@@ -0,0 +1,97 @@
+        $this->maxRetries = $maxRetries;
+
+        if ($delayMilliseconds < 0) {
+            throw new \InvalidArgumentException(sprintf('Delay must be greater than or equal to zero: "%s" passed.', $delayMilliseconds));
```
should be the one from the component, not the root exception

nicolas-grekas reviewed (approved):
The design looks great to me!

Nyholm commented on src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php:
```diff
@@ -32,21 +33,25 @@
 class ConsumeMessagesCommand extends Command
 {
-    protected static $defaultName = 'messenger:consume-messages';
+    protected static $defaultName = 'messenger:consume';
```
You are sneaky adding this change in a large PR. :)

I’m :+1:

Nyholm commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -23,11 +27,31 @@ interface ReceiverInterface
      * Note that this envelope can be `null` if the timeout to receive something has expired.
+     *
+     * If the received message cannot be decoded, the MessageDecodingFailedException
+     * should be caught, the message should be removed from the queue,
```
Using the word “queue” assumes a specific transport. A transport might talk to a different data structure.

Nyholm commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -23,11 +27,31 @@ interface ReceiverInterface
     public function stop(): void;
+
+    /**
+     * Acknowledge (to the queue) that the passed message was handled.
```
Same here.

Nyholm commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -23,11 +27,31 @@ interface ReceiverInterface
+    public function acknowledge(Envelope $envelope): void;
+
+    /**
+     * Called when handling the message failed and it should be removed from the queue.
```
Same here

Nyholm commented on src/Symfony/Component/Messenger/Worker.php:
```diff
@@ -45,10 +69,109 @@ public function run()
+
+                $this->dispatchEvent(new WorkerMessageFailedEvent($envelope, $this->receiverName, $throwable, $shouldRequeue));
+
+                if ($shouldRequeue) {
```
`$shouldBeRetried`

Nyholm reviewed (approved):
I like it. 
We should consider not using the word "queue" other then on AMQP related classes. A Transport is a generic thing.

weaverryan commented:
Rebased, but then added 1 more commit to fix a big. When we retry (and thus, re-send through the entire bus), we need to be smart enough to ONLY send to the transport that the failed message was received from.

For example, if we send a message to `transportA` and `transportB`, if we receive a message from `transportA`, it needs to send back to `transportA` for a retry, but NOT back to `transportB`. If a message is received from an external transport (i.e. it's a message Messenger did not originally send), then it will NOT be retried. Retrying requires us to see a `SentStamp` on the message, which indicates that *we* actually send it originally.

weaverryan commented:
Ready to go again!

Last commits guarantee that redeliveries are only sent back to the same transport.

Nyholm commented:
There are still quite a few mentions of "queue" when we really mean "transport". I think we should try to be more technically correct.

sroze commented on src/Symfony/Component/Messenger/Event/WorkerMessageHandlingEvent.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ * @experimental in 4.3
+ */
+class WorkerMessageHandlingEvent extends AbstractWorkerMessageEvent
```
I'd name it `WorkerMessageReceivedEvent` or `WorkerWillHandleMessageEvent`, otherwise it's not clear enough IMO.

sroze reviewed (commented):
Fabulous. Last thing is the event name I think.

@fabpot can you update your review? 🙏

fabpot commented on src/Symfony/Component/Messenger/Transport/Receiver/ReceiverInterface.php:
```diff
@@ -23,11 +27,31 @@ interface ReceiverInterface
+     *
+     * @throws TransportException If there is an issue communicating with the transport
+     */
+    public function acknowledge(Envelope $envelope): void;
```
Suggestion: rename `acknowledge()` (which is hard to type without a typo, at least for me) to `ack()`?

weaverryan commented:
Last changes look good to me. Thank you!

fabpot commented:
@weaverryan Can you squash so that I can merge? Thank you.

weaverryan commented:
Squashed!

fabpot commented:
Thank you @weaverryan.
